lib/pbench/server/database/models/users.py

```diff
@@ -61,12 +126,11 @@ def query(id=None, username=None, email=None) -> Optional["User"]:
         Returns:
             A User object if a user is found, None otherwise.
         """
+        dbsq = Database.db_session.query(User)
         if id:
-            user = Database.db_session.query(User).filter_by(id=id).first()
+            user = dbsq.filter_by(id=id).first()
         elif username:
-            user = Database.db_session.query(User).filter_by(username=username).first()
-        elif email:
-            user = Database.db_session.query(User).filter_by(email=email).first()
+            user = dbsq.filter_by(username=username).first()
```

Comment:
This leads me to ask:  do we **really** need a mechanism for fetching the `User` value(s) associated with a given _username_?  (If not, we can avoid this problem by simply removing the capability from here.)

I'm wondering, when we ask for a `User` based on a username, if we aren't actually setting up to ask for something else (like all the _datasets_ associated with a username).  If that's the case, then perhaps we can finesse the issue by avoiding the intermediate question and skipping to the final one:  that is, instead of asking for the `User`(s) associated with the username, just as for the `Dataset`'s associated with the username, and then implement the function which produces that result directly, instead of overloading this function here.

Reply:
> That is, if our broker is configured with more than one ID provider, then there can be more than one user with the same username.

I don't think that is possible, even though the identity of the user is provided by the IDP, the user identity federation is internally managed by the Keycloak Broker. Take a look at the [first login flow](https://www.keycloak.org/docs/11.0/server_admin/#_identity_broker_first_login) of Keycloak. How the broker manages the user identity internally is completely up to the broker configuration, and by default the usernames are unique. I don't think it's worth worrying about all these details right now as we don't even know which identity providers the Intlab Keycloak broker is gonna have, It might only have Redhat IDP. At this point, we don't even know about the configurational details of the Keycloak broker so we don't know whether the email/username going to be unique so I wouldn't worry about this until we have a clear idea about our broker. 

Until that if we decide to rely on our implementation of a broker, then I don't think we would have similar usernames situation, we might even provide very limited identity providers like google or github.

Reply:
OK, based on that reference, it sounds like Keycloak itself is imposing the restriction that username and email are unique within the realm, and that if multiple OIDC Id providers validate the same username/email then those external accounts must be linked to a single account within the realm.  (I didn't see an option mentioned for permitting non-unique usernames...the options all seemed to be about how to reconcile them.)

Given that that's the model, that's a relief for us -- that makes things simpler.

Although, it calls into question whether we need or want to deal with UUIDs at all...if we're guaranteed that Keycloak will present us with unique usernames, we can you them to identify users, and we don't need UUIDs...and then we don't need the `Users` table at all...unless usernames are, in fact, mutable....